subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/ResolvedArtifactsGraphVisitor.java

```diff
@@ -65,9 +65,11 @@ public void visitSelector(DependencyGraphSelector selector) {
     @Override
     public void visitEdges(DependencyGraphNode node) {
         for (DependencyGraphEdge dependency : node.getIncomingEdges()) {
-            DependencyGraphNode parent = dependency.getFrom();
-            ArtifactsForNode artifacts = getArtifacts(dependency, node);
-            artifactResults.visitArtifacts(parent, node, artifacts.artifactSetId, artifacts.artifactSet);
+            if (dependency.contributesArtifacts()) {
+                DependencyGraphNode parent = dependency.getFrom();
+                ArtifactsForNode artifacts = getArtifacts(dependency, node);
+                artifactResults.visitArtifacts(parent, node, artifacts.artifactSetId, artifacts.artifactSet);
+            }
```

Comment:
I wonder if the additional artifact visitation per constraint was the source of some of our dependency regression? (This seems like it could be an expensive thing to do for every pending dependency).

Reply:
Looking at the performance results on this branch, I can not see any significant improvement. But maybe we can explore this direction further. So if an edge is pending, we always know some things like:

- the edge never contributes artifacts
- there is always at least one additional (non-pending) incoming edge

Maybe this information can be used earlier during resolution to short-circuit some computation(s).